batcher/aligned-sdk/src/communication/messaging.rs

```diff
@@ -55,178 +55,198 @@ pub async fn send_messages(
+            error!(
+                "Batcher responded with invalid payment service address: {:?}, expected: {:?}",
+                received_addr, expected_addr
             );
+            Err(SubmitError::InvalidPaymentServiceAddress(
+                received_addr,
+                expected_addr,
+            ))
+        }
+        Ok(ResponseMessage::InvalidProof(reason)) => {
+            error!("Batcher responded with invalid proof: {}", reason);
+            Err(SubmitError::InvalidProof(reason))
+        }
+        Ok(ResponseMessage::CreateNewTaskError(merkle_root, error)) => {
+            error!("Batcher responded with create new task error: {}", error);
+            Err(SubmitError::BatchSubmissionFailed(
+                "Could not create task with merkle root ".to_owned()
+                    + &merkle_root
+                    + ", failed with error: "
+                    + &error,
+            ))
         }
         Ok(ResponseMessage::ProtocolVersion(_)) => {
-            return Err(SubmitError::UnexpectedBatcherResponse(
+            error!("Batcher responded with protocol version instead of batch inclusion data");
+            Err(SubmitError::UnexpectedBatcherResponse(
                 "Batcher responded with protocol version instead of batch inclusion data"
                     .to_string(),
-            ));
+            ))
         }
         Ok(ResponseMessage::BatchReset) => {
-            return Err(SubmitError::ProofQueueFlushed);
+            error!("Batcher responded with batch reset");
+            Err(SubmitError::ProofQueueFlushed)
         }
         Ok(ResponseMessage::Error(e)) => {
             error!("Batcher responded with error: {}", e);
-        }
-        Ok(ResponseMessage::CreateNewTaskError(merkle_root)) => {
-            return Err(SubmitError::BatchSubmissionFailed(
-                "Could not create task with merkle root ".to_owned() + &merkle_root,
-            ));
-        }
-        Ok(ResponseMessage::InvalidProof(reason)) => {
-            return Err(SubmitError::InvalidProof(reason));
+            Err(SubmitError::GenericError(e))
         }
         Err(e) => {
-            return Err(SubmitError::SerializationError(e));
+            error!("Error while deserializing batch inclusion data: {}", e);
+            Err(SubmitError::SerializationError(e))
+        }
+    }
+}
+
+// Used to match the message received from the batcher,
+// with the NoncedVerificationData you sent
+// This is used to verify the proof you sent was indeed included in the batch
+fn match_batcher_response_with_stored_verification_data(
+    batch_inclusion_data: &BatchInclusionData,
+    sent_verification_data: &mut Vec<NoncedVerificationData>,
+) -> Result<VerificationDataCommitment, SubmitError> {
+    debug!("Matching verification data with batcher response ...");
+    let mut index = None;
+    for (i, sent_nonced_verification_data) in sent_verification_data.iter_mut().enumerate() {
```

Comment:
sent_verification_data is already sorted by nonce, yes. Also incoming batch_inclusion_data are theoretically also in nonce-order (batcher sends in order of its inclusion in a batch, which is always processed in nonce-order, and all are sent in the same TCP WS so they should always be arriving in the order they were sent). So iterating in this order ends up being more efficient.

Reply:
Anyhow, the swap_remove could be changed to a `if element is the first one, do pop instead of swap_remove` for more efficiency

Reply:
But, if we only care for the last nonce we should iterate the vec in reverse order, filtering any errors until we find one successful submit. That will ensure us all previous proofs were also submitted (if I understood your other comment correctly)

Reply:
I agree with Juli. The last element of `sent_verification_data` is the one with the lowest nonce, so we should iterate over it in reverse order. In that case, we wouldn't need `swap_remove` since removing the last element has the same complexity as `pop()`

Reply:
Actually `.push()` adds at the end of the array , same as `.pop()`. I will reverse the array to avoid `swap_remove` while taking advantage of the low cost of removing from the end of the array.
If i find an easy way to iterate in reverse i will implement it as well